### PR TITLE
Add HTML Repr for `Security` Class

### PR DESCRIPTION
**Pull request by freyam**
- [x] Tests added / passed
- [x] Passes `black distributed` / `flake8 distributed` / `isort distributed`

I have added the HTML Representation of the `Security` class and also tweaked the `__repr__` to specify where the keys/certs are stored.

## `_repr_html_`

image

## `__repr__`

image

/cc @martindurant @GenevieveBuckley

### CODE REVIEW

**Comment by freyam** [comment]
This is a Draft PR as of now as I have not figured out some key visual details in this representation. 

I am working on the following at the moment:
1. Make the strings look cleaner
2. Have some kind of SVG image show up next to the title (a key and a lock perhaps)
3. Add more relevant info
4. Maybe also add information about the client + cluster as well (good idea?)

**Review by jrbourbeau — COMMENTED** [review]
Thanks @freyam! 

I wonder if we want to avoid printing certs and private keys to the screen for security reasons. I see in the normal string repr, we're explicitly replacing those values with "...". If we do end up displaying them, I'll recommend we put them in a `<details>` dropdown in order to keep things more compact and manageable. 

cc @jacobtomlinson @jcrist

**Comment by GenevieveBuckley** [comment]
I'm +1 on using a `<details>` dropdown for any long-ish values.

Perhaps @jacobtomlinson can chime in with what information he'd like to see for debugging, without compromising security. I'd imagine we never want to print private keys. I'm not sure about when or how we'd want to use the certificates, but security is pretty well outside my wheelhouse.

**Comment by jacobtomlinson** [comment]
I just had a sync with @freyam about this (and other things). Main takeaways were:

- Require encryption is useful to know.
- Displaying the keys themselves isn't good because they will get saved into notebooks.
- Knowing whether the object was created using `Security.temporary()` or `Security(**paths_to_keys)` would be really useful. Particularly because with `Security.temporary()` the keys are generated dynamically and the only copy is kept in memory.

**Comment by freyam** [comment]
image
@GenevieveBuckley @jacobtomlinson 

Are you looking for something like this?

**Comment by jacobtomlinson** [comment]
That's looking great! I think the full path in the second one would be helpful though.

**Comment by freyam** [comment]
added :heavy_check_mark: 

image

**Comment by freyam** [comment]
I like "Security". It's simple.

image

**Comment by GenevieveBuckley** [comment]
I like "Security" too. Simple is nice.

This is looking great @freyam

**Comment by GenevieveBuckley** [comment]
I'm not sure why the macos-latest CI test failed, I don't think that has anything to do with this PR

```
 FAILED distributed/tests/test_client.py::test_cancel_multi_client
```

<details>
<summary>Details:</summary>

```
=================================== FAILURES ===================================
___________________________ test_cancel_multi_client ___________________________

s = <Scheduler: "tcp://127.0.0.1:51346" workers: 0 cores: 0, tasks: 0>
a = <Worker: 'tcp://127.0.0.1:51347', 0, Status.closed, stored: 1, running: 0/1, ready: 0, comm: 0, waiting: 0>
b = <Worker: 'tcp://127.0.0.1:51349', 1, Status.closed, stored: 0, running: 0/2, ready: 0, comm: 0, waiting: 0>

    @gen_cluster()
    async def test_cancel_multi_client(s, a, b):
        c = await Client(s.address, asynchronous=True)
        f = await Client(s.address, asynchronous=True)
    
        x = c.submit(slowinc, 1)
        y = f.submit(slowinc, 1)
    
        assert x.key == y.key
    
        await c.cancel([x])
    
        assert x.cancelled()
        assert not y.cancelled()
    
        start = time()
        while y.key not in s.tasks:
            await asyncio.sleep(0.01)
            assert time() < start + 5
    
>       out = await y

distributed/tests/test_client.py:2369: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Future: cancelled, type: builtins.int, key: slowinc-8d878816ee46401dd0f7dff237d708bd>
raiseit = True

    async def _result(self, raiseit=True):
        await self._state.wait()
        if self.status == "error":
            exc = clean_exception(self._state.exception, self._state.traceback)
            if raiseit:
                typ, exc, tb = exc
                raise exc.with_traceback(tb)
            else:
                return exc
        elif self.status == "cancelled":
            exception = CancelledError(self.key)
            if raiseit:
>               raise exception
E               concurrent.futures._base.CancelledError: slowinc-8d878816ee46401dd0f7dff237d708bd

distributed/client.py:251: CancelledError
----------------------------- Captured stderr call -----------------------------
distributed.scheduler - INFO - Clear task state
distributed.scheduler - INFO -   Scheduler at:     tcp://127.0.0.1:51346
distributed.scheduler - INFO -   dashboard at:           127.0.0.1:51345
distributed.worker - INFO -       Start worker at:      tcp://127.0.0.1:51347
distributed.worker - INFO -          Listening to:      tcp://127.0.0.1:51347
distributed.worker - INFO -          dashboard at:            127.0.0.1:51348
distributed.worker - INFO - Waiting to connect to:      tcp://127.0.0.1:51346
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          1
distributed.worker - INFO -                Memory:                  14.00 GiB
distributed.worker - INFO -       Local Directory: /Users/runner/work/distributed/distributed/dask-worker-space/worker-eww7f940
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -       Start worker at:      tcp://127.0.0.1:51349
distributed.worker - INFO -          Listening to:      tcp://127.0.0.1:51349
distributed.worker - INFO -          dashboard at:            127.0.0.1:51350
distributed.worker - INFO - Waiting to connect to:      tcp://127.0.0.1:51346
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          2
distributed.worker - INFO -                Memory:                  14.00 GiB
distributed.worker - INFO -       Local Directory: /Users/runner/work/distributed/distributed/dask-worker-space/worker-uds0p9o5
distributed.worker - INFO - -------------------------------------------------
distributed.scheduler - INFO - Register worker <WorkerState 'tcp://127.0.0.1:51349', name: 1, memory: 0, processing: 0>
distributed.scheduler - INFO - Starting worker compute stream, tcp://127.0.0.1:51349
distributed.core - INFO - Starting established connection
distributed.scheduler - INFO - Register worker <WorkerState 'tcp://127.0.0.1:51347', name: 0, memory: 0, processing: 0>
distributed.scheduler - INFO - Starting worker compute stream, tcp://127.0.0.1:51347
distributed.core - INFO - Starting established connection
distributed.worker - INFO -         Registered to:      tcp://127.0.0.1:51346
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -         Registered to:      tcp://127.0.0.1:51346
distributed.worker - INFO - -------------------------------------------------
distributed.core - INFO - Starting established connection
distributed.core - INFO - Starting established connection
distributed.scheduler - INFO - Receive client connection: Client-531dd1b0-fb97-11eb-8a63-005056881560
distributed.core - INFO - Starting established connection
distributed.scheduler - INFO - Receive client connection: Client-531fdd48-fb97-11eb-8a63-005056881560
distributed.core - INFO - Starting established connection
distributed.scheduler - INFO - Client Client-531dd1b0-fb97-11eb-8a63-005056881560 requests to cancel 1 keys
distributed.scheduler - INFO - Scheduler cancels key slowinc-8d878816ee46401dd0f7dff237d708bd.  Force=False
distributed.worker - INFO - Stopping worker at tcp://127.0.0.1:51347
distributed.worker - INFO - Stopping worker at tcp://127.0.0.1:51349
distributed.scheduler - INFO - Remove worker <WorkerState 'tcp://127.0.0.1:51347', name: 0, memory: 1, processing: 0>
distributed.core - INFO - Removing comms to tcp://127.0.0.1:51347
distributed.scheduler - INFO - Remove worker <WorkerState 'tcp://127.0.0.1:51349', name: 1, memory: 0, processing: 1>
distributed.core - INFO - Removing comms to tcp://127.0.0.1:51349
distributed.scheduler - INFO - Lost all workers
distributed.scheduler - INFO - Scheduler closing...
distributed.scheduler - INFO - Scheduler closing all comms
```

</details>

**Comment by GenevieveBuckley** [comment]
Followup work to this could include updating `__repr__` to add paths to security certificates or "Temporary (in memory)" labels.

**Review by GenevieveBuckley — APPROVED** [review]

**Comment by freyam** [comment]
## Two Options

1. Dict: more readable and clean
image

2. List: more concise
image

@GenevieveBuckley which one is more intuitive?

**Comment by freyam** [comment]
@GenevieveBuckley I have added the list implementation for now to maintain consistency. It used to be in the form of a list before as well. (I like the dict variant more, but depends on which one you prefer)

**I didn't create a new function for the duplicate code as it's not the same.** 

There's a tiny difference (`require_encryption` key is removed in `_repr_html_`) that would make it inefficient to compute in a separate function. So, I just used them individually in their respective methods.

**Comment by freyam** [comment]
# Updates

1. Added the function for the duplicate code: `attr_to_dict()`. Cleaner code :rocket:
2. list implementation as getting errors with the dict implementation (`__repr__` should return a string and not a dict. if I use str(dict), it looks similar to list. So, went with list implementation)

**Comment by freyam** [comment]
## `_repr_html_`

image

## `__repr__`

image

File: distributed/security.py [hunk 1]
```diff
@@ -153,18 +153,57 @@ def _set_field(self, kwargs, field, config_name):
...
-        return "Security(" + ", ".join(f"{k}={v}" for k, v in items) + ")"
+                    attr[k] = val
+
+        return attr
+
+    def __repr__(self):
+        attr = self.attr_to_dict()
+        return (
+            "Security("
+            + ", ".join(f"{key}={value}" for key, value in attr.items())
```
Comment (GenevieveBuckley) [inline distributed/security.py:178]:
```suggestion
            + ",\n    ".join(f"{key}={value}" for key, value in attr.items())
```

What do you think adding newlines to improve reading/scanning the text?

File: distributed/security.py [hunk 2]
```diff
@@ -153,18 +153,57 @@ def _set_field(self, kwargs, field, config_name):
...
+            + ")"
```
Comment (GenevieveBuckley) [inline distributed/security.py:179]:
```suggestion
            + "\n)"
```

(If you try the new line suggestion, we'll want the close bracket also on its own line, so it's easier to see)

File: distributed/security.py (hunk 1 above)
Comment (GenevieveBuckley) [reply distributed/security.py:178]: Oh, I see above you experimented with using a dictionary to get a similar kind of behaviour. Yes, I like the way that looks a lot more - it's much easier to skim through the information and find what you need.

**Comment by GenevieveBuckley** [comment]
> @GenevieveBuckley I have added the list implementation for now to maintain consistency. It used to be in the form of a list before as well. (I like the dict variant more, but depends on which one you prefer)

I like it more too, it's much easier to skim read.

> **I didn't create a new function for the duplicate code as it's not the same.**
> 
> There's a tiny difference (`require_encryption` key is removed in `_repr_html_`) that would make it inefficient to compute in a separate function. So, I just used them individually in their respective methods.

The main point here is that we *want* to have the same content for `__repr__` and `_repr_html_`. Minor differences are a problem, because we assume we'll get the same information, however we choose to look at it.

I see `require_encryption` listed in both representations, so I think this problem is resolved now?

**Review by GenevieveBuckley — APPROVED** [review]
Approved (ideally pending the text representation change discussed above)

**Comment by freyam** [comment]
image
the `\n` shows up on the output as well. 

I searched around to find ways to make this work. No solutions yet.

So, reverting back to the one-line list for now.

File: distributed/security.py [hunk 3]
```diff
@@ -153,18 +153,57 @@ def _set_field(self, kwargs, field, config_name):
...
-        return "Security(" + ", ".join(f"{k}={v}" for k, v in items) + ")"
+                    attr[k] = val
+
+        return attr
+
+    def __repr__(self):
+        attr = self.attr_to_dict()
+        return (
+            "Security("
+            + ", ".join(f"'{key}'='{value}'" for key, value in attr.items())
```
Comment (jacobtomlinson) [inline distributed/security.py:178]: Let's remove the single quotes here. Folks may think things are strings when they are not.

```suggestion
            + ", ".join(f"{key}={value}" for key, value in attr.items())
```

File: distributed/security.py [hunk 4]
```diff
@@ -153,18 +153,57 @@ def _set_field(self, kwargs, field, config_name):
             out = dask.config.get(config_name)
         setattr(self, field, out)
 
-    def __repr__(self):
+    def attr_to_dict(self):
```
Comment (jacobtomlinson) [inline distributed/security.py:156]: This probably wants to be private.

```suggestion
    def _attr_to_dict(self):
```

**Review by jacobtomlinson — COMMENTED** [review]
This is looking awesome.

A couple of last nits but this should be good to merge soon.

File: distributed/security.py (hunk 4 above)
Comment (freyam) [reply distributed/security.py:156]: Oh right ✔️

**Comment by freyam** [comment]
Added ✔️

**Review by jacobtomlinson — APPROVED** [review]

**Comment by GenevieveBuckley** [comment]
Looks like your comments have all been addressed, are you happy to merge now @jacobtomlinson?

**Comment by freyam** [comment]
:yellow_heart:

**Pull request merged**